### PR TITLE
Repeat specification for repeatWhen

### PR DESCRIPTION
**Pull request by Desislav-Petrov**
hi @chemicL and @OlegDokuka - kicked off some work for https://github.com/reactor/reactor-core/issues/3545 but wanted to verify the approach before i go any further.

So my plan is as follows.

1) Work out a base approach to get the simple repeat test pass
2) Add necessary functionality for the rest of the tests to pass
3) Refactor and pull the common bits between repeat/retry as previously suggested
4) See if anything else is missing in the repeat functionality and add it

### CODE REVIEW

File: reactor-core/src/main/java/reactor/util/repeat/RepeatSpec.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public Flux<Long> generateCompanion(Flux<RepeatSignal> other) {
+        return Flux.deferContextual(cv ->
+                other
+                        .contextWrite(cv)
+                        .concatMap(retryWhenState -> {
+                            RepeatSignal copy = retryWhenState.copy();
+                            long iteration = copy.getRepeatsSoFar();
+                            if (iteration > repeats) {
```
Comment (Desislav-Petrov) [inline reactor-core/src/main/java/reactor/util/repeat/RepeatSpec.java:31]: I don't think this is the most graceful way to do this, was trying to get some inspiration from the Retry one but the case is slightly different there

**Comment by chemicL** [comment]
Hey, @Desislav-Petrov. I'm trying to go through some PRs that were not given enough attention. Please accept the team's apologies. I had a brief look and would like to know whether after not gaining much attention from us you are still willing to make progress on this.

My first concern is a public API change - we need to make sure that anything that we add does not remove existing functionality.

Also, the next(-1) signal looks suspicious - in case of differences between the implementation in reactor-extras, can you list them with a high-level explanation?

Let me know, thanks.

**Comment by Desislav-Petrov** [comment]
Hey @chemicL - I'm still keen on working on that as long as I get some support/gudaince - let me take a look since it's been a while to summarise the idea.

**Comment by chemicL** [comment]
@Desislav-Petrov take your time. I'll try to be helpful. Should I be occupied to the limits with other concerns, I'll be sure to let you know.

**Comment by chemicL** [comment]
@Desislav-Petrov it's been a month since my last message. Let's reopen if you think you'll find the time to make some progress.

**Pull request closed without merging**